src/os/bluestore/BlueStore.cc

```diff
@@ -9121,7 +9121,182 @@ int BlueStore::_split_collection(TransContext *txc,
   return r;
 }
 
+#define KEY_SLAB 32
+#define VALUE_SLAB 64
 
+int get_key_slab(size_t sz)
+{
+  return (sz/KEY_SLAB);
+}
+
+string get_key_slab_to_range(int slab)
+{
+  int lower_bound = slab * KEY_SLAB;
+  int upper_bound = (slab + 1) * KEY_SLAB;
+  string ret = "[" + stringify(lower_bound) + "," + stringify(upper_bound) + ")";
+  return ret;
+}
+
+int get_value_slab(size_t sz)
+{
+  return (sz/VALUE_SLAB);
+}
+
+string get_value_slab_to_range(int slab)
+{
+  int lower_bound = slab * VALUE_SLAB;
+  int upper_bound = (slab + 1) * VALUE_SLAB;
+  string ret = "[" + stringify(lower_bound) + "," + stringify(upper_bound) + ")";
+  return ret;
+}
+
+//histogram
+struct value_dist
+{
+  uint64_t count;
+  uint32_t max_len;
+};
+
+struct key_dist
+{
+  uint64_t count;
+  uint32_t max_len;
+  map<int, struct value_dist> val_map; // slab id to count and others
+};
 
+map<string, map<int, struct key_dist> > key_hist;
+map<int, uint64_t> value_hist; // only value distribution histogram
```

Comment:
Can we put the histogram bits into a class? Should be able to include update_hist_entry?